### PR TITLE
Some fixes

### PR DESCRIPTION
**Pull request by mauropasse**
Some fixes for bugs which didn't appeared yet during testing, but could happen.

- Protect `on_new_event_callback_` with mutex before being read.
- Declare QoS event callback before event handle, to avoid possible data races when an event is destroyed while it's callback is being called.

### CODE REVIEW

**Comment by mauropasse** [comment]
To verify: Do we need mutex on `event_handlers_` ? Currently not protected.

https://github.com/irobot-ros/rclcpp/blob/fef3728ac1c47b50f2e9485c33c66f90d34e03f8/rclcpp/include/rclcpp/publisher_base.hpp#L339-L340

Applies also for `subscription_base.hpp` 

https://github.com/irobot-ros/rclcpp/blob/fef3728ac1c47b50f2e9485c33c66f90d34e03f8/rclcpp/include/rclcpp/subscription_base.hpp#L573-L574

**Comment by mauropasse** [comment]
Also this commit could be applied to add safety to `QOSEventHandlerBase` events
https://github.com/mauropasse/rmw_fastrtps/commit/adf1aa7e52b9aea8c1b76886683ae5e663237b81

File: rclcpp/src/rclcpp/qos_event.cpp [hunk 1]
```diff
@@ -35,6 +35,8 @@ UnsupportedEventTypeException::UnsupportedEventTypeException(
 
 QOSEventHandlerBase::~QOSEventHandlerBase()
 {
+  std::lock_guard<std::recursive_mutex> lock(callback_mutex_);
```
Comment (alsora) [inline rclcpp/src/rclcpp/qos_event.cpp:38]: `clear_on_ready_callback` already checks internally whether we have a callback or not (and it has a mutex), so here we can just call it directly (without checks and without mutex)

**Comment by alsora** [comment]
Please open a PR also to the official rclcpp for this fix

**Comment by alsora** [comment]
Sorry for hijacking your PR, but I'm about to do a new Humble update so I wanted to include it.
I replaced this PR with https://github.com/irobot-ros/rclcpp/pull/103

**Pull request closed without merging**